### PR TITLE
test(annotation): cover redistribute

### PR DESCRIPTION
**Pull request by UkiStefanovic**
In this pull request, the function redistribute(...) from the /distribution/main.py module is covered. The tests for this function can be found in the file /tests/test_distribution.py

### CODE REVIEW

File: annotation/tests/test_distribution.py [hunk 1]
```diff
@@ -1516,3 +1524,144 @@ def test_prepare_response():
...
+            file_id=1,
+            pages=list(range(20)),
+        ),
+    ]
+
+    mock_db = Mock(spec=Session)
+    mock_query = mock_db.query.return_value
+    mock_filter = mock_query.filter.return_value
+    mock_all = mock_filter.all
+    mock_all.return_value = task_list
```
Comment (Negoje) [inline annotation/tests/test_distribution.py:1552]: I think you can merge this into one line no need to put it into multiple lines

**Review by Negoje — COMMENTED** [review]
Everything else seems good to me

File: annotation/tests/test_distribution.py (hunk 1 above)
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py]: Yes, it can be shortened. It'll be fixed

File: annotation/tests/test_distribution.py [hunk 2]
```diff
@@ -1516,3 +1523,139 @@ def test_prepare_response():
             )
             == expected_output
         )
+
+
+@pytest.fixture
+def mock_session_and_tasks():
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1529]: Any block of code should do only one thing. Need to mock session? Session fixture. Need to mock tasks? Tasks fixture

https://realpython.com/solid-principles-python/

File: annotation/tests/test_distribution.py [hunk 3]
```diff
@@ -1516,3 +1523,139 @@ def test_prepare_response():
...
+            status=TaskStatusEnumSchema.finished,
+            is_validation=False,
+            file_id=1,
+            pages=list(range(20)),
+        ),
+    ]
+
+    mock_db = Mock(spec=Session, flush=Mock(), rollback=Mock())
+    mock_db.query.return_value.filter.return_value.all.return_value = task_list
+
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1549]: excessive blank lines

File: annotation/tests/test_distribution.py [hunk 4]
```diff
@@ -1516,3 +1523,139 @@ def test_prepare_response():
...
+    mock_db, task_list = mock_session_and_tasks
+    (
+        mock_get_pages,
+        mock_distribute,
+        mock_set_task_statuses,
+    ) = patch_redistribute_dependencies
+
+    def delete_tasks_side_effect(_, tasks_to_delete):
+        nonlocal task_list
+        task_list = [task for task in task_list if task not in tasks_to_delete]
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1605]: I dont see the point of this side effect. Remove it and change the patch to make it simplier.

File: annotation/tests/test_distribution.py [hunk 5]
```diff
@@ -1516,3 +1523,139 @@ def test_prepare_response():
...
+    with patch(
+        "annotation.distribution.main.delete_tasks",
+        side_effect=delete_tasks_side_effect,
+    ) as mock_delete_tasks, patch(
+        "annotation.distribution.main.update_job_status", return_value=None
+    ):
+        redistribute(mock_db, "dummy_token", "x_tenant", job)
+        if job.status == "in_progress":
+            mock_set_task_statuses.assert_called_once_with(job, task_list)
+        mock_delete_tasks.assert_called_once_with(mock_db, set())
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1616]: Simplify this, please

File: annotation/tests/test_distribution.py [hunk 6]
```diff
@@ -1516,3 +1523,139 @@ def test_prepare_response():
...
+):
+    # creating job object
+    job_id = 0
+    callback_url = "url"
+    job = Mock(
+        job_id=job_id,
+        files=[],
+        status=JobStatusEnumSchema.in_progress,
+        callback_url=callback_url,
+    )
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1630]: create a real job object, no need for mock in this place

File: annotation/tests/test_distribution.py [hunk 7]
```diff
@@ -1516,3 +1523,139 @@ def test_prepare_response():
...
+        redistribute(mock_db, "dummy_token", "x_tenant", job)
+        if job.status == "in_progress":
+            mock_set_task_statuses.assert_called_once_with(job, task_list)
+        mock_delete_tasks.assert_called_once_with(mock_db, set())
+
+
+def test_redistribute_error(
+    mock_session_and_tasks: Mock, patch_redistribute_dependencies: Mock
+):
+    # creating job object
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1622]: code is self explanatory, remove this comment, please

File: annotation/tests/test_distribution.py [hunk 8]
```diff
@@ -1516,3 +1523,139 @@ def test_prepare_response():
...
+    mock_db, task_list = mock_session_and_tasks
+    (
+        mock_get_pages,
+        mock_distribute,
+        mock_set_task_statuses,
+    ) = patch_redistribute_dependencies
+
+    def delete_tasks_side_effect(_, tasks_to_delete):
+        nonlocal task_list
+        task_list = [task for task in task_list if task not in tasks_to_delete]
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1641]: same as previous comment  - too complex, adjust your patches

File: annotation/tests/test_distribution.py [hunk 9]
```diff
@@ -1516,3 +1523,139 @@ def test_prepare_response():
...
+        "annotation.distribution.main.delete_tasks",
+        side_effect=delete_tasks_side_effect,
+    ), patch(
+        "annotation.distribution.main.update_job_status",
+        side_effect=JobUpdateException("Connection timeout"),
+    ) as mock_update_job_status, pytest.raises(
+        HTTPException
+    ) as exc_info:
+        redistribute(mock_db, "dummy_token", "x_tenant", job)
+
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1653]:
```suggestion
```

File: annotation/tests/test_distribution.py (hunk 4 above)
Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py:1609]: My idea was to have a mock database with tasks. This function `delete_tasks` is supposed to delete tasks from the database, but i thought i would manipulate the list of tasks i pass directly. I'm not sure how to make this idea simpler

**Comment by UkiStefanovic** [comment]
The pr checks didn't pass, but before I fix them, I would like to see how we could simplify the patch for delete_tasks

File: annotation/tests/test_distribution.py [hunk 10]
```diff
@@ -28,17 +30,64 @@
...
+            status=TaskStatusEnumSchema.finished,
+            is_validation=False,
+            file_id=1,
+            pages=list(range(20)),
+        ),
+    )
+
+
+@pytest.fixture
+def mock_session_redistribute(tasks_for_test_redistribute):
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:71]: typings

File: annotation/tests/test_distribution.py [hunk 11]
```diff
@@ -28,17 +30,64 @@
...
+@pytest.fixture
+def patch_redistribute_dependencies():
+    with patch(
+        "annotation.distribution.main.get_pages_in_work", return_value=None
+    ) as mock_get_pages, patch(
+        "annotation.distribution.main.distribute", return_value=None
+    ) as mock_distribute, patch(
+        "annotation.distribution.main.set_task_statuses"
+    ) as mock_set_task_statuses:
+        yield mock_get_pages, mock_distribute, mock_set_task_statuses
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:88]: simplify this, please. It is too complex

File: annotation/tests/test_distribution.py [hunk 12]
```diff
@@ -1516,3 +1565,111 @@ def test_prepare_response():
...
+            ],
+            JobStatusEnumSchema.in_progress,
+        ),
+        ([], JobStatusEnumSchema.finished),
+    ),
+)
+def test_redistribute(
+    tasks_for_test_redistribute,
+    mock_session_redistribute,
+    patch_redistribute_dependencies,
```
Comment (cakeinsauce) [inline annotation/tests/test_distribution.py:1589]: Typings

File: annotation/tests/test_distribution.py (hunk 4 above)
Comment (cakeinsauce) [reply annotation/tests/test_distribution.py]: But why do you need that? you dont check for values it changes really, you check if it as called with certain params, that's all. You dont need such side effect

Comment (UkiStefanovic) [reply annotation/tests/test_distribution.py]: Yeah, I get it now. I mocked the function with a side effect but only now realize I didn't need any side effect. I'll remove that

**Review by UkiStefanovic — COMMENTED** [review]
Fixed the commented code

**Comment by UkiStefanovic** [comment]
I merged the contents of this pr with the pr 916, so this one is no longer relevant and ill close it

**Pull request closed without merging**